events/oio_events_queue_beanstalkd.c

```diff
@@ -151,88 +158,135 @@ _send (int fd, struct iovec *iov, unsigned int iovcount)
 
-	if (!_is_success((gchar*) buf))
-		return NEWERROR(CODE_BAD_REQUEST, "reply error: unexpected");
+	dst[r] = 0;
 	return NULL;
 }
 
 static GError *
-_put (int fd, gchar *msg, size_t msglen)
+_put_job (int fd, gchar *msg, size_t msglen)
 {
 	gchar buf[256];
-	struct iovec iov[3];
 
+	/* send the request */
 	gsize len = g_snprintf (buf, sizeof(buf),
 			"put %u %u %u %"G_GSIZE_FORMAT"\r\n",
 			oio_events_beanstalkd_default_prio,
 			(guint) oio_events_beanstalkd_default_delay,
 			(guint) oio_events_beanstalkd_default_ttr,
 			msglen);
+	struct iovec iov[3] = {};
 	iov[0].iov_base = buf;
 	iov[0].iov_len = len;
 	iov[1].iov_base = msg;
 	iov[1].iov_len = msglen;
 	iov[2].iov_base = "\r\n";
 	iov[2].iov_len = 2;
+	GError *err = _send_and_read_reply(fd, iov, 3, buf, sizeof(buf));
+	if (err) return err;
+
+	/* No need to retry, the event has been saved ... or explicitely
+	 * dropped. */
+	const char * const replies_ok[] = { "INSERTED", "BURIED", "DRAINING", NULL };
+	for (const char * const *pmsg = replies_ok; *pmsg ;++pmsg) {
+		if (g_str_has_prefix(buf, *pmsg))
+			return NULL;
+	}
 
-	return _send_and_read_reply (fd, iov, 3);
+	return _match_common_error(buf);
 }
 
 static GError *
 _use_tube (int fd, const char *name)
 {
+	gchar buf[256];
+
 	if (!oio_str_is_set(name))
 		return NULL;
 
-	gchar buf[256];
+	/* senf the request */
 	gsize len = g_snprintf(buf, sizeof(buf), "use %s\r\n", name);
-
 	if (len >= sizeof(buf))
-		return NEWERROR(CODE_INTERNAL_ERROR, "BUG: tube name too long");
-
-	struct iovec iov[1];
+		return SYSERR("BUG: tube name too long");
+	struct iovec iov[1] = {};
 	iov[0].iov_base = buf;
 	iov[0].iov_len = len;
-	return _send_and_read_reply (fd, iov, 1);
+	GError *err = _send_and_read_reply(fd, iov, 1, buf, sizeof(buf));
+	if (err) return err;
+
+	/* parse the reply */
+	if (g_str_has_prefix(buf, "USING"))
+		return NULL;
+	return _match_common_error(buf);
+}
+
+static GError *
+_check_server(int fd)
+{
+	gchar buf[2048];
+
+	/* send the request */
+	strcpy(buf, "stats\r\n");
+	struct iovec iov[1] = {};
+	iov[0].iov_base = buf;
+	iov[0].iov_len = strlen(buf);
+	GError *err = _send_and_read_reply(fd, iov, 1, buf, sizeof(buf));
+	if (err) return err;
+
+	/* parse the reply */
+	if (!g_str_has_prefix(buf, "OK"))
+		return _match_common_error(buf);
+
+	gint64 total = 0;
+	gchar **lines = g_strsplit(buf, "\n", -1);
+	if (lines) {
+		for (gchar **pline = lines; *pline ;++pline) {
+			gchar *line = *pline;
+			gint64 count = 0;
+			if (!g_str_has_prefix(line, "current-jobs-"))
+				continue;
+			if (!(line = strchr(line, ':')))
+				continue;
+			if (!oio_str_is_number(g_strchug(line+1), &count))
+				continue;
+			total += count;
+		}
+		g_strfreev(lines);
+	}
+
+	const gint64 max_jobs = oio_events_beanstalkd_check_level_deny;
+	if (max_jobs > 0 && total > max_jobs)
+		return BUSY("FULL");
+	const gint64 alert = oio_events_beanstalkd_check_level_alert;
+	if (alert > 0 && total > alert)
+		GRID_WARN("Beanstalkd too loaded");
```

Comment:
You should print the alert and error limits in the warning (and the error message, since it will also trigger a warning trace).